File: integration_tests/rust/src/degenerate.rs
use {
    crate::{
        assertions::assert_round_trip,
        types::degenerate::types::{EmptyChoiceIn, EmptyChoiceOut, EmptyStructIn, EmptyStructOut},
    },
    std::io,
};

#[allow(clippy::needless_pass_by_value)]
fn _initial_in<T>(x: EmptyChoiceIn) -> T {
    match x {}
}

#[allow(clippy::needless_pass_by_value)]
fn _initial_out<T>(x: EmptyChoiceOut) -> T {
    match x {}
}

fn _terminal_in<T>(_: T) -> EmptyStructIn {
    EmptyStructIn {}
}

fn _terminal_out<T>(_: T) -> EmptyStructOut {
    EmptyStructOut {}
}

pub fn run() -> io::Result<()> {
    assert_round_trip::<EmptyStructOut, EmptyStructIn>(&EmptyStructOut {})?;

    Ok(())
}

File: integration_tests/typescript-node/src/degenerate.ts
import { Degenerate } from '../generated/types';
import { assertRoundTrip } from './assertions';

// eslint-disable-next-line @typescript-eslint/no-unused-vars -- We only care that it type checks.
function initialIn<T>(x: Degenerate.Types.EmptyChoiceIn): T {
  return x;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars -- We only care that it type checks.
function initialOut<T>(x: Degenerate.Types.EmptyChoiceOut): T {
  return x;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars -- We only care that it type checks.
function terminalIn<T>(x: T): Degenerate.Types.EmptyStructIn {
  return {};
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars -- We only care that it type checks.
function terminalOut<T>(x: T): Degenerate.Types.EmptyStructOut {
  return {};
}

export default function run(): void {
  assertRoundTrip(
    Degenerate.Types.EmptyStruct.size,
    Degenerate.Types.EmptyStruct.serialize,
    Degenerate.Types.EmptyStruct.deserialize,
    {},
  );
}
